File: phone/src/bridge/WatchConnector.ts
/**
 * WatchConnector - DSoftBus integration for Watch 5 communication
 * Subscribes to sensor data stream from the wearable device
 */

// import { DSoftBusManager } from '@kit.DistributedServiceKit'; // Actual HarmonyOS SDK import

export interface SensorPayload {
  heartRate: number;
  accelerometer: [number, number, number];
  timestamp: number;
}

export interface PetStateUpdate {
  mood: string;
  energy: number;
  timestamp: number;
}

export class WatchConnector {
  private isConnected: boolean = false;
  private dataCallback?: (payload: SensorPayload) => void;
  private static readonly WATCH_DEVICE_ID = "TamagotchiWatch";

  /**
   * Initialize DSoftBus and connect to Watch 5
   */
  connect(): void {
    console.log("WatchConnector: Initializing DSoftBus connection...");

    // TODO: Implement DSoftBus connection
    // DSoftBusManager.initialize();
    // DSoftBusManager.subscribe(WatchConnector.WATCH_DEVICE_ID, (rawData) => {
    //     const payload = this.parseSensorData(rawData);
    //     this.dataCallback?.(payload);
    // });

    this.isConnected = true;
    console.log("WatchConnector: Connected to Watch 5");
  }

  /**
   * Subscribe to sensor data updates
   * @param callback Function to call when new data arrives
   */
  onDataReceived(callback: (payload: SensorPayload) => void): void {
    this.dataCallback = callback;
  }

  /**
   * Send pet state update back to watch
   * @param petState Updated pet state from cloud
   */
  sendPetStateUpdate(petState: PetStateUpdate): void {
    if (!this.isConnected) {
      console.log("WatchConnector: Not connected, cannot send update");
      return;
    }

    // TODO: Implement DSoftBus publish
    // DSoftBusManager.publish(WatchConnector.WATCH_DEVICE_ID, JSON.stringify(petState));
    console.log(`WatchConnector: Sent pet state update - ${petState.mood}`);
  }

  /**
   * Disconnect from watch
   */
  disconnect(): void {
    // TODO: Cleanup DSoftBus resources
    // DSoftBusManager.unsubscribe(WatchConnector.WATCH_DEVICE_ID);
    this.isConnected = false;
    console.log("WatchConnector: Disconnected");
  }

  /**
   * Parse raw DSoftBus data into structured payload
   */
  private parseSensorData(rawData: ArrayBuffer): SensorPayload {
    // TODO: Implement actual parsing logic
    // This is a placeholder
    return {
      heartRate: 0,
      accelerometer: [0.0, 0.0, 0.0],
      timestamp: Date.now()
    };
  }
}

File: phone/src/bridge/WatchConnector.kt
/**
 * WatchConnector - DSoftBus integration for Watch 5 communication
 * Subscribes to sensor data stream from the wearable device
 */

// import com.huawei.softbus.DSoftBusManager // Placeholder - actual SDK import

class WatchConnector {
    private var isConnected = false
    private var dataCallback: ((SensorPayload) -> Unit)? = null

    /**
     * Initialize DSoftBus and connect to Watch 5
     */
    fun connect() {
        println("WatchConnector: Initializing DSoftBus connection...")

        // TODO: Implement DSoftBus connection
        // DSoftBusManager.initialize()
        // DSoftBusManager.subscribe(WATCH_DEVICE_ID) { rawData ->
        //     val payload = parseSensorData(rawData)
        //     dataCallback?.invoke(payload)
        // }

        isConnected = true
        println("WatchConnector: Connected to Watch 5")
    }

    /**
     * Subscribe to sensor data updates
     * @param callback Function to call when new data arrives
     */
    fun onDataReceived(callback: (SensorPayload) -> Unit) {
        this.dataCallback = callback
    }

    /**
     * Send pet state update back to watch
     * @param petState Updated pet state from cloud
     */
    fun sendPetStateUpdate(petState: PetStateUpdate) {
        if (!isConnected) {
            println("WatchConnector: Not connected, cannot send update")
            return
        }

        // TODO: Implement DSoftBus publish
        // DSoftBusManager.publish(WATCH_DEVICE_ID, petState.toByteArray())
        println("WatchConnector: Sent pet state update - ${petState.mood}")
    }

    /**
     * Disconnect from watch
     */
    fun disconnect() {
        // TODO: Cleanup DSoftBus resources
        // DSoftBusManager.unsubscribe(WATCH_DEVICE_ID)
        isConnected = false
        println("WatchConnector: Disconnected")
    }

    /**
     * Parse raw DSoftBus data into structured payload
     */
    private fun parseSensorData(rawData: ByteArray): SensorPayload {
        // TODO: Implement actual parsing logic
        // This is a placeholder
        return SensorPayload(
            heartRate = 0,
            accelerometer = Triple(0.0, 0.0, 0.0),
            timestamp = System.currentTimeMillis()
        )
    }

    companion object {
        private const val WATCH_DEVICE_ID = "TamagotchiWatch"
    }
}

// Placeholder data classes (will be moved to shared/contracts)
data class SensorPayload(
    val heartRate: Int,
    val accelerometer: Triple<Double, Double, Double>,
    val timestamp: Long
)

data class PetStateUpdate(
    val mood: String,
    val energy: Int,
    val timestamp: Long
)
